test/functional/test_runner.py

```diff
@@ -529,33 +529,35 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
 
     max_len_name = len(max(test_list, key=len))
     test_count = len(test_list)
-    for i in range(test_count):
-        test_result, testdir, stdout, stderr = job_queue.get_next()
-        test_results.append(test_result)
-        done_str = "{}/{} - {}{}{}".format(i + 1, test_count, BOLD[1], test_result.name, BOLD[0])
-        if test_result.status == "Passed":
-            logging.debug("%s passed, Duration: %s s" % (done_str, test_result.time))
-        elif test_result.status == "Skipped":
-            logging.debug("%s skipped" % (done_str))
-        else:
-            print("%s failed, Duration: %s s\n" % (done_str, test_result.time))
-            print(BOLD[1] + 'stdout:\n' + BOLD[0] + stdout + '\n')
-            print(BOLD[1] + 'stderr:\n' + BOLD[0] + stderr + '\n')
-            if combined_logs_len and os.path.isdir(testdir):
-                # Print the final `combinedlogslen` lines of the combined logs
-                print('{}Combine the logs and print the last {} lines ...{}'.format(BOLD[1], combined_logs_len, BOLD[0]))
-                print('\n============')
-                print('{}Combined log for {}:{}'.format(BOLD[1], testdir, BOLD[0]))
-                print('============\n')
-                combined_logs_args = [sys.executable, os.path.join(tests_dir, 'combine_logs.py'), testdir]
-                if BOLD[0]:
-                    combined_logs_args += ['--color']
-                combined_logs, _ = subprocess.Popen(combined_logs_args, universal_newlines=True, stdout=subprocess.PIPE).communicate()
-                print("\n".join(deque(combined_logs.splitlines(), combined_logs_len)))
-
-            if failfast:
-                logging.debug("Early exiting after test failure")
-                break
+    i = 0
+    while i < test_count:
+        for test_result, testdir, stdout, stderr in job_queue.get_next():
```

Comment:
```
...............
               
Remaining jobs: [wallet_import_rescan.py --legacy-wallet, p2p_node_network_limited.py]
..............................................................................................................................................................................................................................
                                                                                                                                                                                                                              
Remaining jobs: [wallet_import_rescan.py --legacy-wallet]
......................................................................................................................................................................................
                                                                                                                                                                                      
----------------------------------------------------------------------
Ran 10 tests in 0.753s
OK
Traceback (most recent call last):
  File "test/functional/test_runner.py", line 816, in <module>
    main()
  File "test/functional/test_runner.py", line 460, in main
    run_tests(
  File "test/functional/test_runner.py", line 535, in run_tests
    for test_result, testdir, stdout, stderr in job_queue.get_next():
  File "test/functional/test_runner.py", line 647, in get_next
    raise IndexError('pop from empty list')
IndexError: pop from empty list
```

https://cirrus-ci.com/task/6633293054476288?logs=ci#L6037

Reply:
Hmm, I don't understand how this is possible. It requires `get_next()` to be called when `self.jobs` and `self.test_list` are empty, while `i < test_count` in `run_tests()`. The latter should imply that `self.test_list` is not empty.

Reply:
See https://github.com/bitcoin/bitcoin/pull/23995

Reply:
see https://github.com/bitcoin/bitcoin/pull/23995#issuecomment-1007436418 for a possible explanation.